merkle-tree/hash-set/src/lib.rs

```diff
@@ -0,0 +1,872 @@
+
+        let offset = offset
+            + (mem::size_of::<Option<I>>()
+                * usize::try_from(hash_set_ref.capacity_indices).unwrap());
+        hash_set_ref.values =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<HashSetCell>).unwrap();
+
+        hash_set_ref
+    }
+
+    /// Casts a byte slice into `HashSet` and then initializes it.
+    ///
+    /// * `bytes` is casted into a reference of `HashSet` and used as
+    ///   storage for the struct.
+    /// * `capacity_indices` indicates the size of the indices table. It should
+    ///   already include a desired load factor and be greater than the expected
+    ///   number of elements to avoid filling the set too early and avoid
+    ///   creating clusters.
+    /// * `capacity_values` indicates the size of the values array. It should be
+    ///   equal to the number of expected elements, without load factor.
+    /// * `sequence_threshold` indicates a difference of sequence numbers which
+    ///   make elements of the has set expired. Expiration means that they can
+    ///   be replaced during insertion of new elements with sequence numbers
+    ///   higher by at least a threshold.
+    ///
+    /// # Purpose
+    ///
+    /// This method is meant to be used mostly in Solana programs to initialize
+    /// a new account which is supposed to store the hash set.
+    ///
+    /// # Safety
+    ///
+    /// This is highly unsafe. Ensuring the alignment and that the slice has
+    /// a correct size, which is able to fit the hash set, is the caller's
+    /// responsibility.
+    ///
+    /// Calling it in async context (or anywhere where the underlying data can
+    /// be moved in memory) is certainly going to cause undefined behavior.
+    pub unsafe fn from_bytes_init(
+        bytes: &mut [u8],
+        capacity_indices: I,
+        capacity_values: I,
+        sequence_threshold: usize,
+    ) -> &mut Self {
+        let hash_set_ref = &mut *(bytes.as_mut_ptr() as *mut Self);
+
+        hash_set_ref.capacity_indices = capacity_indices;
+        hash_set_ref.capacity_values = capacity_values;
+        hash_set_ref.sequence_threshold = sequence_threshold;
+
+        let offset = mem::size_of::<Self>();
+        hash_set_ref.indices =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<I>).unwrap();
+
+        let offset = offset
+            + (mem::size_of::<Option<I>>()
+                * usize::try_from(hash_set_ref.capacity_indices).unwrap());
+        hash_set_ref.values =
+            NonNull::new(bytes.as_mut_ptr().add(offset) as *mut Option<HashSetCell>).unwrap();
+
+        hash_set_ref
+    }
+
+    // Insert a value into the hash set
+    pub fn insert(&mut self, value: &BigUint, sequence_number: usize) -> Result<(), HashSetError> {
+        for i in
+            0..usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?
```

Comment:
```suggestion
        let limit = usize::try_from(self.capacity_values).map_err(|_| HashSetError::IntegerOverflow)?;
        for i in
            0..limit
```


Then it`s more obvious what`s going on here.

Reply:
I decided to store `capacity_values` as usize. It's just one field for the whole hash set and it spare us doing these convertions.

Now I'm only converting usize to u16 (generic `I` type) when writing to indices array.